Add a Websocket server that allows for interaction with running queries

RobinTF opened the pull request:
When posting a query to QLever via a blocking HTTP request, it is now possible to assign a unique ID to this query. Using this ID it is possible to also open a Websocket connection at the same time that will receive status updates on the running query. Currently, the partial runtime information trees are sent after each operation has finished. In the future, we will always get complete trees, and the websocket mechanism can also be used to interact with the query, for example to cancel it.

RobinTF commented on CMakeLists.txt:
```diff
@@ -111,7 +111,7 @@ endif()
 ######################################
 # BOOST
 ######################################
-find_package(Boost 1.74 COMPONENTS iostreams program_options REQUIRED)
```
I'm not sure if this change is actually required. However 1.80 of beast did have a bug that made the application crash

RobinTF replied:
Easier said than done. There are no prebuilt binaries which means with this change along the build will take 10+ minutes longer when building all of boost. Of course this can be slightly optimized by only compiling the libraries that qlever needs but no compilation will always be faster than any compilation

RobinTF commented on src/engine/Server.cpp:
```diff
@@ -469,6 +469,23 @@ nlohmann::json Server::composeCacheStatsJson() const {
+  }
+  auto queryId = OwningQueryId::uniqueIdFromString(std::string(queryIdHeader));
+  if (!queryId) {
+    throw std::runtime_error{"QueryId '" + queryIdHeader +
```
This currently results in a 400 Bad Request reponse. Not sure if it's appropriate in this case

RobinTF replied:
I think 409 would be appropriate here. According to [RFC9110](https://www.rfc-editor.org/rfc/rfc9110#status.409):
> The 409 (Conflict) status code indicates that the request could not be completed due to a conflict with the current state of the target resource. This code is used in situations where the user might be able to resolve the conflict and resubmit the request. The server SHOULD generate content that includes enough information for a user to recognize the source of the conflict.
>
> Conflicts are most likely to occur in response to a PUT request. For example, if versioning were being used and the representation being PUT included changes to a resource that conflict with those made by an earlier (third-party) request, the origin server might use a 409 response to indicate that it can't complete the request. In this case, the response representation would likely contain information useful for merging the differences based on the revision history.

400 status codes usually indicate syntactic or logical errors with the request per [RFC9110](https://httpwg.org/specs/rfc9110.html#status.400):
> The 400 (Bad Request) status code indicates that the server cannot or will not process the request due to something that is perceived to be a client error (e.g., malformed request syntax, invalid request message framing, or deceptive request routing).

RobinTF commented on src/util/http/beast.h:
```diff
@@ -18,11 +18,17 @@
 // Needed for libc++ in C++20 mode, because std::result_of was removed.
+#ifndef BOOST_ASIO_HAS_STD_INVOKE_RESULT
 #define BOOST_ASIO_HAS_STD_INVOKE_RESULT
+#endif
```
These changes are just to please some redefinition warnings

RobinTF commented on src/util/websocket/Common.h:
```diff
@@ -0,0 +1,108 @@
+};
+
+namespace ad_utility::websocket::common {
+class OwningQueryId {
```
This is a wrapper for `QueryId` which in practice is owned by the query execution context and destroyed along with it

RobinTF commented on src/util/websocket/WebSocketManager.cpp:
```diff
@@ -0,0 +1,241 @@
+  // Generate unique webSocketIds until it overflows and hopefully the old
+  // ids are gone by then.
+  static WebSocketId uniqueId() {
+    static std::atomic<uint32_t> webSocketIdCounter{0};
```
As discussed we might want to consider an address based id assignment system

joka921 replied:
We definitely want to do this. 
Just make sure, that the address is stable. But do you need the address at all, or can you just pass the callbacks directly? I will see.

RobinTF commented on src/util/websocket/WebSocketManager.cpp:
```diff
@@ -0,0 +1,241 @@
+  }
+
+  // In case an unexpected exception is thrown, close the connection gracefully.
+  if (ws.is_open()) {
```
I'm not sure if it is worth trying to catch all kind of exceptions here. If the TCP socket is closed by force the websocket connection will end either way

joka921 commented on src/util/websocket/Common.h:
```diff
@@ -0,0 +1,108 @@
+    })());
+  }
+
+ public:
+  // Prevent copies to prevent preemptive cleanup
+  OwningQueryId(const OwningQueryId&) = delete;
+  OwningQueryId& operator=(const OwningQueryId&) = delete;
+
+  OwningQueryId(OwningQueryId&&) = default;
+  OwningQueryId& operator=(OwningQueryId&&) = default;
+
+  static std::optional<OwningQueryId> uniqueIdFromString(std::string id) {
+    auto queryId = QueryId::idFromString(std::move(id));
+    std::lock_guard lock{registryMutex_};
+    if (registry_.count(queryId)) {
+      return std::nullopt;
+    }
+    registry_.insert(queryId);
+    return OwningQueryId{std::move(queryId)};
+  }
+
+  static OwningQueryId uniqueId() {
+    static thread_local std::mt19937 generator(std::random_device{}());
+    std::uniform_int_distribution<uint64_t> distrib{};
+    std::optional<OwningQueryId> result = std::nullopt;
+    while (!result.has_value()) {
+      result = uniqueIdFromString(std::to_string(distrib(generator)));
+    }
+    return std::move(result.value());
+  }
+
+  const QueryId& toQueryId() const noexcept { return id_; }
+
+  ~OwningQueryId() {
+    if (!id_.empty()) {
+      std::lock_guard lock{registryMutex_};
+      registry_.erase(id_);
+    }
+  }
+};
```
I understand the idea, but a more idiomatic interface (especially without `static` stuff is the following):
Use a 
`QueryIdManager` class, that owns the registry in a threadsafe way.
The various functions `createNewQueryId` and `registerQueryId` return a 
`std::unique_ptr<QueryId, CustomDeleterThatUnregistersTheQueryId>`. 
These should be `[[nodiscard]]`
The synchronization of the register is best done with a `ad_utility::Synchronized` to couple the mutex and the data it is protecting.

joka921 replied:
(Then you should also be able to get rid of the `recursive_mutex`.

joka921 commented on src/util/websocket/Common.h:
```diff
@@ -0,0 +1,108 @@
+
+using TimeStamp = std::chrono::time_point<std::chrono::steady_clock>;
+
+struct TimedClientPayloadStruct {
+  std::string payload;
+  TimeStamp updateMoment;
+};
```
Just call this `PayloadAndTimeStamp` or `DataAndTimestamp`.

joka921 commented on src/util/websocket/QueryState.cpp:
```diff
@@ -0,0 +1,47 @@
+  auto snapshot = std::make_shared<TimedClientPayload::element_type>(
+      std::move(payload), std::chrono::steady_clock::now());
+  // Only insert data if there are additional waiting websockets
+  if (websocket::fireAllCallbacksForQuery(queryId, snapshot)) {
+    std::lock_guard lock{queryStatesMutex};
```
That is an advantage of having a more complex `QueryCallBackManagementToken` (bad name) instead of the `QueryId`, then we can more easily handle the locking on a per query basis, and not globaally.

joka921 commented on src/util/websocket/QueryState.cpp:
```diff
@@ -0,0 +1,47 @@
+TimedClientPayload getIfUpdatedSince(const QueryId& queryId,
+                                     websocket::common::TimeStamp timeStamp) {
+  std::lock_guard lock{queryStatesMutex};
+  if (queryStates.count(queryId)) {
```
In general (no matter if this code stays), we have `contains` now for `hash_map` and similar types.

joka921 commented on src/util/websocket/WebSocketManager.cpp:
```diff
@@ -0,0 +1,241 @@
+#include <mutex>
+#include <optional>
+
+#include "../MultiMap.h"
+#include "../http/HttpUtils.h"
+#include "./Common.h"
+#include "./QueryState.h"
+
```
absolute includes please.

joka921 commented on src/util/websocket/WebSocketManager.cpp:
```diff
@@ -0,0 +1,241 @@
+
+}  // namespace ad_utility::websocket
+
+template <>
+struct std::hash<ad_utility::websocket::WebSocketId> {
+  auto operator()(
+      const ad_utility::websocket::WebSocketId& webSocketId) const noexcept {
+    return std::hash<uint32_t>{}(webSocketId.id_);
+  }
+};
```
Same about hashing (if needed at all).

joka921 commented on src/util/websocket/WebSocketManager.cpp:
```diff
@@ -0,0 +1,241 @@
+using QueryUpdateCallback = std::function<void(TimedClientPayload)>;
+
+static std::mutex activeWebSocketsMutex{};
+static absl::btree_multimap<QueryId, WebSocketId> activeWebSockets{};
```
I don't think we need a multimap,
just map from `QueryId` to the `QueryIdToWebsocketIdCallbackManager` object, that is visible from the outside
(This is probably also the object, that the `QueryIdManager` should return and register.

joka921 commented on src/util/websocket/WebSocketManager.cpp:
```diff
@@ -0,0 +1,241 @@
+  std::lock_guard lock1{activeWebSocketsMutex};
+  // make sure websocket has not been shut down right before acquiring the lock
+  if (containsKeyValuePair(activeWebSockets, queryId, webSocketId)) {
+    std::lock_guard lock2{listeningWebSocketsMutex};
+    // Ensure last callback has been fired, otherwise co_await will wait
+    // indefinitely
+    AD_CORRECTNESS_CHECK(listeningWebSockets.count(webSocketId) == 0);
+    listeningWebSockets[webSocketId] = std::move(callback);
```
I don't quite understand this yet, but the two mutexes look a bit suspicious to me. 
Maybe I'll have a look at the `Websocket-coroutine` and then an idea comes to mind.

joka921 commented on src/util/websocket/WebSocketManager.cpp:
```diff
@@ -0,0 +1,241 @@
+}
+
+// ONLY CALL THIS IF YOU HOLD the listeningWebSocketsMutex!!!
+void fireCallbackAndRemoveIfPresentNoLock(WebSocketId webSocketId,
+                                          TimedClientPayload payload) {
+  if (listeningWebSockets.count(webSocketId) != 0) {
+    listeningWebSockets.at(webSocketId)(std::move(payload));
+    listeningWebSockets.erase(webSocketId);
+  }
+}
+
+void fireCallbackAndRemoveIfPresent(WebSocketId webSocketId,
+                                    TimedClientPayload payload) {
+  std::lock_guard lock{listeningWebSocketsMutex};
+  fireCallbackAndRemoveIfPresentNoLock(webSocketId, std::move(payload));
+}
```
The `NoLock` is a reason to make such things classes etc.

RobinTF replied:
In theory you could of course move the "no lock" definition to the header so it can be accessed elsewhere, and making the whole thing private would prevent that, but obviously this doesn't solve the issue that you could accidentally call the wrong function or am I missing something?

joka921 reviewed (commented):
A very partial review of some of the code with rather high-level
architecture suggestions.
I still have to think about the point where we change from `boost::awaitable` to the `mutex` world with the `active` vs `listening` websockets, depending on their position inside the coroutine.
In general, I would aim for less global locks, but mostly keep the synchronization on a per queryId or a per Websocket connection level. Maybe we could even implement the code, 
that sends an update to all websockets for a given queryId also inside the `Boost::Asio` framework, to move this code also towards the server threads.
But again, I'll also have to think about how to best deal with websockets that are currently still writing.

RobinTF commented on src/engine/QueryExecutionContext.h:
```diff
@@ -23,6 +23,8 @@
 #include <string>
 #include <vector>
 
+#include "../util/websocket/Common.h"
```
Regarding import rules. In this header there are quite a lot of imports using the `#include <>` syntax. Is there any difference between these two or is this just inconsistent behaviour?

RobinTF commented on src/util/websocket/Common.h:
```diff
@@ -0,0 +1,108 @@
+};
+}  // namespace ad_utility::websocket::common
+
+template <>
+struct std::hash<ad_utility::websocket::common::QueryId> {
+  auto operator()(
+      const ad_utility::websocket::common::QueryId& queryId) const noexcept {
+    return std::hash<std::string>{}(queryId.id_);
+  }
+};
```
Just to compute the hash or to completely opt into the absl hash mechanism by implementing AbslHashValue?

RobinTF commented:
> A very partial review of some of the code with rather high-level architecture suggestions. I still have to think about the point where we change from `boost::awaitable` to the `mutex` world with the `active` vs `listening` websockets, depending on their position inside the coroutine. In general, I would aim for less global locks, but mostly keep the synchronization on a per queryId or a per Websocket connection level. Maybe we could even implement the code, that sends an update to all websockets for a given queryId also inside the `Boost::Asio` framework, to move this code also towards the server threads. But again, I'll also have to think about how to best deal with websockets that are currently still writing.

The underlying reason for this decision has to do with ownership of the websocket. Currently the websocket is exclusively owned by the coroutine that creates it and everything else is wired around it: The coroutine asks for updates, and waits until it is notified by another thread, Nothing can happen to this instance of websocket outside of the coroutine, it is strictly encapsuled.

Another option would be to share the associated strand and expose some interface allowing to operate on the websocket via callbacks asynchronously. This approach also requires us to use a lock because the strands need to be shared, added and removed across threads. This approach conceptually makes the strand the owner of the websocket object. If the strands lifetime and the lifetime of the websocket are directly coupled. In my opinion however this just introduces another layer on top of the system: Instead of synchronizing access to the broadcasted data this synchronizes access to individual sockets/strands to then trigger asynchronous function calls on them (which are then synchronized by asio using said strands). Also I'm not sure if boost asio gives any guarantees on the order of execution when using co_spawn, or if we had to ensure the ordering is correct ourselves.

I can't think of any other variants right now, there are probably some more but I can't think of any other variant where the lifetime of the websocket object is as clearly defined as here.

RobinTF commented on src/util/websocket/WebSocketManager.cpp:
```diff
@@ -0,0 +1,177 @@
+                   currentNode_ = informationQueues_.at(queryId_).getHead();
+                   (*sharedSelf)(currentNode_->getPayload());
+                 } else {
+                   (*sharedSelf)(nullptr);
```
This is where I'm really unsure about the behaviour of boost asio:
When calling `sharedSelf` here, does this mean the co_await will resume on the `registryStrand_` or on the initial strand it was invoked on? Not sure how to check this, but the former behaviour would be pretty bad in terms of performance

joka921 reviewed (changes requested):
I have posted a review on various levels:
1. Make the code as simple and readable as possible s.t. we can reason about the inherently complicated parts (locking, strands, coroutines, you name it).
2.  Start writing unit tests that are as readable as possible, especially for the corner cases (finishing of queries races with websockets that still need updates etc).  This will be quite some work, but it also helps to understand the asynchronous stuff much better.

joka921 reviewed (changes requested):
I have posted a review on various levels:
1. Make the code as simple and readable as possible s.t. we can reason about the inherently complicated parts (locking, strands, coroutines, you name it).
2.  Start writing unit tests that are as readable as possible, especially for the corner cases (finishing of queries races with websockets that still need updates etc).  This will be quite some work, but it also helps to understand the asynchronous stuff much better.

RobinTF commented on src/util/LinkedList.h:
```diff
@@ -0,0 +1,56 @@
+  }
+
+  std::shared_ptr<Node<type>> getHead() { return head_; }
+};
```
The current approach has the main benefit that it is cleaned up incrementally. So if a query finished it would only keep the last couple of nodes that have not been sent yet. So if this "additional memory requirement" is acceptable it can be changed.

RobinTF commented on src/util/websocket/Common.h:
```diff
@@ -0,0 +1,98 @@
+  explicit QueryId(std::string id) : id_{std::move(id)} {}
+
+ public:
+  static QueryId idFromString(std::string id) { return QueryId{std::move(id)}; }
```
This needs to be public, so `WebSocketManager` can create an instance from the URL the observing client used. This would conceptually a non-owning reference because it lives independently of the actual lifetime of the query

RobinTF commented on src/util/websocket/WebSocketManager.cpp:
```diff
@@ -0,0 +1,177 @@
+               } else {
+                 if (informationQueues_.contains(queryId_)) {
+                   currentNode_ = informationQueues_.at(queryId_).getHead();
+                   (*sharedSelf)(currentNode_->getPayload());
```
I don't see the benefit with this proposal. How would the websockets know if a certain query is finished or not? At some point you have to sort of synchronize the access and if you do that via `informationQueues_` or by some other data structure shouldn't make a difference performance wise?

RobinTF commented on src/util/http/HttpServer.h:
```diff
@@ -216,13 +222,26 @@ class HttpServer {
-        // by QLever's timeout mechanism.
-        stream.expires_never();
+        if (beast::websocket::is_upgrade(req)) {
+          auto errorResponse = ad_utility::websocket::checkPathIsValid(req);
```
Not sure how you imagine the member function to look like. The main reason for this code construct is because of the conditional move in the if. If the websocket manager is not responsible nothing is done and nothing is moved anywhere. But if it is reponsible, the ansible cleanup is canceled and the request/socket stuff is moved to the manager.
Do you see a better approach to this?

RobinTF commented on src/engine/QueryExecutionContext.h:
```diff
@@ -112,16 +119,22 @@ class QueryExecutionContext {
+        nlohmann::ordered_json(runtimeInformation).dump());
+  }
+
+  ~QueryExecutionContext() {
+    webSocketManager_.releaseQuery(owningQueryId_.toQueryId());
+  }
```
Hmm currently the `OwningQueryId` class is completely independent of a websocket manager. Not sure if it would be the right move to change that. So conceptually I thought the `QueryExecutionContext` would be the right place to perform this kind of cleanup because here all the components come together.

RobinTF commented on src/engine/QueryExecutionContext.h:
```diff
@@ -112,16 +119,22 @@ class QueryExecutionContext {
   ad_utility::AllocatorWithLimit<Id> getAllocator() { return _allocator; }
 
+  void signalQueryUpdate(const RuntimeInformation& runtimeInformation) {
+    webSocketManager_.addQueryStatusUpdate(
+        owningQueryId_.toQueryId(),
+        nlohmann::ordered_json(runtimeInformation).dump());
+  }
```
See https://github.com/ad-freiburg/qlever/pull/970#discussion_r1207144981

RobinTF commented on src/engine/QueryExecutionContext.h:
```diff
@@ -112,16 +118,22 @@ class QueryExecutionContext {
+  }
+
+  ~QueryExecutionContext() {
+    webSocketManager_.releaseQuery(owningQueryId_.toQueryId());
```
This is currently a little bit dangerous. If the webSocketManager has no io context the destructor throws

RobinTF commented on src/engine/Server.cpp:
```diff
@@ -578,6 +613,11 @@ boost::asio::awaitable<void> Server::processQuery(
+        getQueryId(request), webSocketManager};
+    auto notifierFunction = [&webSocketNotifier](std::string json) {
+      webSocketNotifier(std::move(json));
+    };
```
@joka921 I was thinking about making this an implicit conversion, but in the end I decided against it because this way you can clearly see that a "conversion" just provides a different view on the whole thing via reference. Thoughts?

RobinTF commented on src/util/CleanupDeleter.h:
```diff
@@ -0,0 +1,41 @@
+  /// obj is about to be deleted.
+  static std::unique_ptr<T, CleanupDeleter<T>> cleanUpAfterUse(
+      std::remove_reference_t<T> obj, std::function<void(type&)> function) {
+    return {new T(std::move(obj)), CleanupDeleter{std::move(function)}};
```
@joka921 While `std::unique_ptr` provides a nice part of this kind of interface it still requires a lot of boilerplate code to make it work, so I'm starting to believe a custom solution using a boolean might be the cleaner approach for this, while still keeping the abstraction layer. Thoughts?

joka921 commented on src/util/http/websocket/WebSocketManager.cpp:
```diff
@@ -0,0 +1,189 @@
+
+    co_await ws.async_accept(request, boost::asio::use_awaitable);
+
+    auto strand = net::make_strand(ws.get_executor());
```
Ah, there is the synchronization....

joka921 commented on src/util/http/websocket/WebSocketManager.cpp:
```diff
@@ -0,0 +1,189 @@
+
+net::awaitable<void> WebSocketManager::manageConnection(
+    tcp::socket socket, http::request<http::string_body> request) {
+  auto executor = socket.get_executor();
+  return net::co_spawn(
+      executor, connectionLifecycle(std::move(socket), std::move(request)),
+      net::use_awaitable);
```
I think the thread sanitizer only likes this, if the socket is either on a strand or on a dedicated `coro_executor` for reasons.

joka921 reviewed (changes requested):
Okay, I am up to speed now and am beginning to understand what is going on there.
Most of the suggestions are minor details, have a look at everything (not all of the comments are in logical order), and then maybe we can set up a meeting for open questions.

joka921 commented on src/util/http/websocket/EphemeralWaitingList.cpp:
```diff
@@ -0,0 +1,68 @@
+
+    if (cancellationSlot.is_connected()) {
+      cancellationSlot.assign(
+          [this, &queryId, &functionId,
+           executor = std::move(executor)](net::cancellation_type) {
+            // Make sure to remove the entry on the same executor
```
Is it really safe to pass in the `queryId` and the `functionId` by reference into that lambda that might be called who knows when. (I am pretty sure that at least the `functionId` is definitely hanging, and the `queryId` is rather dangerous.

RobinTF replied:
The queryId's lifetime in this case is tied to the lifetime of the coroutine handling the websocket request, so it's definitely still alive while said coroutine is still in the process of being cancelled, or rather a sub-awaitable within the call stack of said coroutine (which can catch the cancellation and proceed normally)

joka921 commented on src/util/http/websocket/EphemeralWaitingList.cpp:
```diff
@@ -0,0 +1,68 @@
+           executor = std::move(executor)](net::cancellation_type) {
+            // Make sure to remove the entry on the same executor
+            // which was used to initiate this
+            net::dispatch(executor, [this, &queryId, &functionId]() {
+              removeCallback(queryId, functionId);
+            });
```
Or maybe even better:

Move the data structure of the `Websockets waiting on a particular query` into the `QueryToWebsocketDistribtor` or whereever. Then you have the whole lifetime of a single query (not yet started, running, finished) all in one data structure, and the global manager only has a flat `HashMap<QueryId, shared_ptr<Distributor>` with the invariant that all the currently running queries are guaranteed to be in there.

joka921 replied:
(In the meantime maybe we have a little but too many classes.

joka921 commented on src/util/http/websocket/EphemeralWaitingList.cpp:
```diff
@@ -0,0 +1,68 @@
+  return net::async_initiate<CompletionToken, void()>(
+      initiate, std::forward<CompletionToken>(token));
+}
+
```
// _____________________________________________________________________________________

(Everywhere in the `.cpp` files.

joka921 commented on src/util/http/websocket/UpdateFetcher.h:
```diff
@@ -0,0 +1,48 @@
+
+  WebSocketTracker& webSocketTracker_;
+  const QueryId& queryId_;
+  StrandType& socketStrand_;
+  std::shared_ptr<QueryToSocketDistributor> distributor_{nullptr};
+  // Counter to ensure sequential processing
+  size_t nextIndex_ = 0;
+
+ public:
+  UpdateFetcher(const QueryId& queryId, WebSocketTracker& webSocketTracker,
+                StrandType& strand)
+      : webSocketTracker_{webSocketTracker},
+        queryId_{queryId},
+        socketStrand_{strand} {}
+
+  /// If an update occurred for the given query since the last time this was
+  /// called this resumes immediately. Otherwise it will wait
```
This class should be split up:

The actual `updateFetcher` should only be the `index` and the `sharedPtr<Distributor>`

Then the `WebSocketTracker` should have a function
`awaitable<UpdateFetcher> waitForQueryStart(QueryId)`.

And then this `UpdateFetcher` should either manage its own strand (not passed in), or be not-threadsafe (like the waiting list) and implicitly reuse the strand of the `webSocket` (probably the latter is enough, and similar to what you are currently doing anyway.).
And if that is done, Maybe the remaining functionality goes back into the class that handles a single websocket session, because we are down to one function again here.

joka921 commented on src/util/http/websocket/WebSocketManager.cpp:
```diff
@@ -146,11 +71,13 @@ net::awaitable<void> WebSocketManager::connectionLifecycle(
-        net::co_spawn(strand, waitForServerEvents(ws, queryId),
-                      net::use_awaitable) &&
-        net::co_spawn(strand, handleClientCommands(ws), net::use_awaitable));
+    co_await net::dispatch(strand, net::use_awaitable);
```
I currently don't like that the `closing of the websocket` is the implicit signal for `normal termination`.
I think the semantics that we want is
`wait until one of them either finishes or fails, and then close the websocket if still possible. But it might be, that implementing this would also not lead to nicer code.

joka921 reviewed (changes requested):
I am slowly beginning to understand what is going on here.
In the meantime we have quite a lot of classes, maybe we can consolidate the design again, I have made some suggestions.

RobinTF commented on src/engine/Server.cpp:
```diff
@@ -613,11 +616,10 @@ boost::asio::awaitable<void> Server::processQuery(
-    };
+    AD_CORRECTNESS_CHECK(webSocketTracker_ != nullptr);
+    auto webSocketNotifier =
+        co_await ad_utility::websocket::WebSocketNotifier::create(
+            getQueryId(request), *webSocketTracker_);
```
Conceptually the `WebSocketNotifier` class is supposed to be the bridge between the asio code (io related stuff) and the non-asio code (query logic), so it's intentionally designed this way. However we might be able to change something about the WSTracker & WSManager hierarchy to make it nicer to use. Let's talk about this

RobinTF commented on src/util/http/HttpServer.h:
```diff
@@ -256,8 +260,10 @@ class HttpServer {
+          // This is the error "The socket was closed due to a timeout" or
+          // indicates a socket has been closed mid-operation
+          if (error.code() == beast::error::timeout ||
+              error.code() == boost::asio::error::operation_aborted) {
```
Yes, this error occurs if the websocket is closed while the server is still sending payloads to it

RobinTF commented on src/util/http/websocket/Common.h:
```diff
@@ -26,7 +26,10 @@ class QueryId {
-  bool empty() const noexcept { return id_.empty(); }
+
+  /// Checks if the id is empty. Because empty ids are not allowed,
+  /// this is usually a good indicator if the object has been moved out of
```
The latter case applies, this is only used for sanity checks

RobinTF commented on src/util/http/websocket/Common.h:
```diff
@@ -74,18 +82,20 @@ class QueryRegistry {
+            // Registry might be
+            if (auto registry = weakRegistry.lock()) {
+              registry->wlock()->erase(qid);
+            }
```
Regarding the comment, that was already fixed as we discussed beforehand.

And it's not about if we want to support letting the `QueryId`s outlive the registry, but if we don't the API is just not safe to use:
```c++
QueryRegistry registry{};
some_async_operation(registry.uniqueId())
// oops registry is destructed before the async operation could be completed -> segfault
```

We can't force a movable object to have a fixed lifetime.

RobinTF commented on src/util/http/websocket/EphemeralWaitingList.h:
```diff
@@ -0,0 +1,72 @@
+  /// is called.
+  template <typename CompletionToken>
+  net::awaitable<void> registerCallbackAndWait(const QueryId& queryId,
+                                               CompletionToken&& token);
```
We only use the awaitable case, this can be hardcoded it it's desired.

RobinTF commented on src/util/http/websocket/EphemeralWaitingList.h:
```diff
@@ -0,0 +1,72 @@
+  };
+
+  /// Multimap that maps query ids to multiple callbacks of waiting websockets.
+  std::unordered_multimap<QueryId, IdentifiableFunction, absl::Hash<QueryId>>
+      waitingCallbacks_{};
```
Removing single objects doesn't required an additional check if the wrapping vector needs to be destroyed for multimaps

RobinTF commented on src/util/http/websocket/QueryToSocketDistributor.cpp:
```diff
@@ -0,0 +1,77 @@
+    if (cancellationSlot.is_connected()) {
+      cancellationSlot.assign([this](net::cancellation_type) {
+        // Make sure wakeupCalls_ is accessed safely.
+        net::dispatch(strand_, [this]() { wakeupCalls_.clear(); });
+      });
+    }
+  };
```
Nothing to explain, this is just a faulty implementation 😄

RobinTF commented on src/util/http/websocket/WebSocketTracker.cpp:
```diff
@@ -0,0 +1,37 @@
+  if (socketDistributors_.contains(queryId)) {
+    co_return socketDistributors_.at(queryId);
+  } else {
+    co_await waitingList_.waitForQueryStart(queryId);
+    // The same strand used in createDistributor, but just to be sure
+    co_await net::dispatch(globalStrand_, net::use_awaitable);
+    co_return socketDistributors_.at(queryId);
```
So in the grand scheme of things this class _does_ provide this kind of interface for callers of _this_ class. The alternative would be to merge this class with `EphemeralWaitingList`, which would then allow do do this here

RobinTF commented on src/util/http/websocket/WebSocketTracker.h:
```diff
@@ -0,0 +1,49 @@
+/// unique for this class, so in the common case of this class being used
+/// globally the provided thread-safety comes at a cost.
+class WebSocketTracker {
+  net::io_context& ioContext_;
```
Yes, because we need it to create new strands for every `QueryId`

RobinTF commented on src/util/http/websocket/WebSocketNotifier.h:
```diff
@@ -5,29 +5,47 @@
 
-  /// Notifies all listeners the query was completed.
-  ~WebSocketNotifier();
+  /// Asynchronously creates an instance of this class. This is because because
+  /// creating a distributor for this class needs to be done in a synchronized
+  /// way.
+  static net::awaitable<WebSocketNotifier> create(
+      common::OwningQueryId owningQueryId, WebSocketTracker& webSocketTracker);
```
See https://github.com/ad-freiburg/qlever/pull/970#discussion_r1315788757

RobinTF commented on src/util/http/websocket/WebSocketNotifier.cpp:
```diff
@@ -2,15 +2,34 @@
-WebSocketNotifier::~WebSocketNotifier() {
-  webSocketManager_.releaseQuery(owningQueryId_.toQueryId());
+void WebSocketNotifier::operator()(std::string json) const {
+  net::co_spawn(
+      executor_,
+      [distributor = distributor_,
+       json = std::move(json)]() mutable -> net::awaitable<void> {
+        co_await distributor->addQueryStatusUpdate(std::move(json));
+      },
+      net::detached);
```
If `co_spawn` were in the `distributor` this would be undefined behaviour in the following case:
1. `distributor->addQueryStatusUpdate()` is called
2. distributor shared ptr is destructed, incrementing the count to 0 and destroying the object
3. async operation is run on a deleted object
4. pray for segfault instead of weird stuff

This is also why a copy of the shared pointer is added as closure for the lambda. The alternative would be to pass the distributor a shared pointer of itself which is arguably terrible API design

RobinTF commented on src/util/http/websocket/UpdateFetcher.cpp:
```diff
@@ -0,0 +1,24 @@
+namespace ad_utility::websocket {
+
+net::awaitable<UpdateFetcher::PayloadType> UpdateFetcher::waitForEvent() {
+  co_await net::dispatch(socketStrand_, net::use_awaitable);
```
Valid point

RobinTF commented on src/util/UniqueCleanup.h:
```diff
@@ -0,0 +1,70 @@
+
+  ~UniqueCleanup() {
+    if (active_) {
+      function_(std::move(value_));
```
Currently the interface makes it very explicit that you _could_ move the object somewhere else just before destruction if you wanted to. If you don't make use of that, it will be destroyed afterwards anyways

RobinTF commented on src/util/http/websocket/EphemeralWaitingList.h:
```diff
@@ -0,0 +1,71 @@
+
+  /// Helper struct to bundle a std::function with a unique id
+  struct IdentifiableFunction {
+    std::function<void()> func_;
+    FunctionId id_;
```
Currently the callbacks are wrapped in lambdas anyways, so we might be able to template the type

joka921 commented on src/util/Asio.h:
```diff
@@ -0,0 +1,25 @@
+#ifndef QLEVER_ASIO_H
+#define QLEVER_ASIO_H
+
+#include <boost/asio/awaitable.hpp>
```
This file should have a somewhat longer name like `BoostAsioUtils` or `AsioHelpers` etc.

joka921 commented on src/util/Asio.h:
```diff
@@ -0,0 +1,25 @@
+namespace net = boost::asio;
+
+/// Helper function that ensures that co_await resumes on the same
+/// executor it started with.
```
Do you also need the utility 
`onExecutor(awaitableOrFunction, executor)`:

* dispatch to the `executor`
* run the `awaitableOrFunction`
* resume to the original executor?

RobinTF replied:
I believe there are not enough places right now where this would be providing actual benefit. Also because there's no "decorator" mechanism in C++ like there's in python, this would require to introduce another lambda or function just to wrap it which doesn't help readability. I'd prefer a simple "co_await" guard for the correct strand at the start of the function

joka921 commented on src/util/UniqueCleanup.h:
```diff
@@ -0,0 +1,70 @@
+
+/// Wrapper class that allows to call a function
+/// just before the wrapper value T is destroyed.
+template <std::move_constructible T>
```
If you want, you can still do the templating, but for your acutal use-cases you then have to template on `std::function<...>`.

joka921 commented on src/util/http/websocket/QueryHub.cpp:
```diff
@@ -0,0 +1,53 @@
+  auto coroutine = [](auto strand, auto& distributors, auto queryId,
+                      auto id) -> net::awaitable<void> {
+    co_await net::dispatch(strand, net::use_awaitable);
+    // Only erase object if we created it, otherwise we have a race condition
+    if (distributors.contains(queryId) && distributors.at(queryId).id_ == id) {
```
Do you need the additional `id` or can you solve this via the address of the distributor that is to be destroyed directly? (I am not sure if this works with weak pointers). Its address should still be valid during destruction, but please check.
I now overall like the synchronization, but it is complicated (nature of things). Especially the usage of the `weak_ptr` as part of avoiding ABA-problems is nice, but very subtle.

joka921 commented on src/util/http/websocket/QueryHub.cpp:
```diff
@@ -0,0 +1,53 @@
+  }
+
+  auto id = counter++;
+  auto coroutine = [](auto strand, auto& distributors, auto queryId,
```
Should be called cleanupAction or `unregisterDistributorFromHub` or something and be commented.

joka921 commented on src/util/http/websocket/QueryHub.cpp:
```diff
@@ -0,0 +1,53 @@
+      ioContext_,
+      [this, queryId, coroutine = std::move(coroutine), id]() mutable {
+        net::co_spawn(globalStrand_,
+                      // TODO check if references are safe here?
```
They are safe as long as the `QueryHub` outlives everything IMHO, does that answer your question?

joka921 commented on src/util/http/websocket/QueryToSocketDistributor.h:
```diff
@@ -9,28 +9,43 @@
   /// Flag to indicate if a query ended and won't receive any more updates.
   bool finished_ = false;
 
+  /// Function to remove this distributor from the registry when it is
```
The `registry` is now called `QueryHub`.

joka921 commented on src/util/http/websocket/QueryToSocketDistributor.cpp:
```diff
@@ -7,52 +7,78 @@
 template <typename CompletionToken>
 auto QueryToSocketDistributor::waitForUpdate(CompletionToken&& token) {
-  auto initiate = [this]<typename Handler>(Handler&& self) mutable {
+  auto id = counter++;
```
How is this threadsafe when the `counter` is not atomic?

RobinTF replied:
`waitForUpdate()` is private API and only ever called on the same strand

joka921 commented on src/util/http/websocket/UpdateWrapper.h:
```diff
@@ -0,0 +1,57 @@
+                       };
+                       net::co_spawn(
+                           executor,
+                           coroutine(std::forward<decltype(distributor)>(
+                               distributor)),
```
Can use the `AD_FWD()` macro for `std::forward` in most cases.

joka921 reviewed (commented):
I think we are converging on the design.
You can definitely start writing tests, especially two kinds:
* Simple tests that document the usage pattern and dependencies
* Tests with many threads that make it probable to detect race conditions.

joka921 commented on test/WebSocketSessionTest.cpp:
```diff
@@ -0,0 +1,159 @@
+  tcp::socket client{ioContext};
+  co_await connect(server, client);
+
+  auto controllerActions = [&]() -> net::awaitable<void> {
```
can this be called `clientLogic`or `clientActions`?

RobinTF replied:
`verifySessionEndsWhenServerIsDoneSending` does trigger some server related actions, so it wouldn't be accurate in that case

joka921 commented on test/WebSocketSessionTest.cpp:
```diff
@@ -0,0 +1,159 @@
+// Hack to allow ASSERT_*() macros to work with ASYNC_TEST
+#define return co_return
+
+ASYNC_TEST(WebSocketSession, verifySessionEndsOnClientCloseWhileTransmitting) {
```
I don't quite understand the name. Why `while transmitting`?

RobinTF replied:
While updates are transmitted from the server to the client, before they are completed

RobinTF replied:
In other words it tests that the server handles an abrupt socket close gracefully

joka921 commented on test/WebSocketSessionTest.cpp:
```diff
@@ -0,0 +1,159 @@
+
+  co_await (serverLogic() && controllerActions());
+}
+
```
In general (maybe have a pass over your tests)
we also add the `// ______________________________________________________________` lines before all test cases to make them visually stand out.

joka921 reviewed (changes requested):
In general this looks now very nice. 
With the naming we are slowly converging.

In general I think that the `QueryToSocketDistributor` class should be hidden as much as possible, as the safe interfaces to use the facilities are the `UpdateFetcher` and the `UpdateWrapper`. So probably the `QueryHub` should directly hand out those two classes, and the Distributor should only be private and friends to those classes.

RobinTF commented on src/util/http/websocket/QueryHub.h:
```diff
@@ -22,30 +21,26 @@ using StrandType = net::strand<net::any_io_executor>;
 
   /// Creates a new `QueryToSocketDistributor` or returns the pre-existing
-  /// for the provided query id if there already is one.
+  /// for the provided query id if there already is one. The bool parameter
+  /// should be set to true if the caller of this methods intends to use
+  /// the distributor object to send data, rather than just receiving it.
```
It's important to understand that `true` indicates a new query has started, which automatically invalidates any older running queries. So yes, it's perfectly safe and could theoretically occur in production if there's really heavy load in my opinion (haven't reproduced in in the real world, but the tests cover this case).

RobinTF commented on src/util/http/websocket/QueryToSocketDistributor.cpp:
```diff
@@ -4,52 +4,33 @@
+  } catch (boost::system::system_error& error) {
+    if (error.code() != net::error::operation_aborted) {
+      throw;
     }
```
It's possible to avoid the try catch block in this case, but when using coroutines the errors are propagated via exceptions, so rethrowing is pretty simple instead of having to rebuild a system error using the acquired error code. I'd suggest we discuss this a small change for a future PR, because there are several ways to do it and this PR is already big enough

RobinTF commented on src/util/http/websocket/UpdateWrapper.h:
```diff
@@ -0,0 +1,54 @@
+        executor_{std::move(executor)} {}
+
+ public:
+  UpdateWrapper(UpdateWrapper&&) noexcept = default;
```
No

RobinTF commented on src/util/http/websocket/WebSocketSession.h:
```diff
@@ -0,0 +1,55 @@
+using websocket::QueryId;
+using websocket = beast::websocket::stream<tcp::socket>;
+
+/// Class to manage the lifecycle of a single websocket. Single-use only.
```
This is a left-over from the time where the API could be called multiple times with the same object. No longer needed

RobinTF commented on src/util/http/websocket/WebSocketSession.h:
```diff
@@ -0,0 +1,55 @@
+  UpdateFetcher updateFetcher_;
+  websocket ws_;
+
+  /// Loop that waits for input from the client
```
Yes, but that's not happening _yet_ 😬

RobinTF commented on src/util/http/websocket/WebSocketSession.cpp:
```diff
@@ -0,0 +1,116 @@
+    auto json = co_await updateFetcher_.waitForEvent();
+    if (json == nullptr) {
+      if (ws_.is_open()) {
+        co_await ws_.async_close(beast::websocket::close_code::normal,
+                                 net::use_awaitable);
+      }
+      break;
+    }
```
That would be an option too

RobinTF commented on src/util/http/websocket/WebSocketSession.cpp:
```diff
@@ -0,0 +1,116 @@
+      co_return;
+    }
+    // There was an unexpected error, rethrow
+    throw;
```
Yes, this will get caught by the big catch block that catches exceptions during every TCP session

RobinTF commented on test/util/AsyncTestHelpers.h:
```diff
@@ -0,0 +1,44 @@
+
+template <typename Func>
+void runCoroutine(Func innerRun) {
+  net::io_context ioContext{};
+
+  auto future = net::co_spawn(ioContext, innerRun(ioContext), net::use_future);
+  ioContext.run_for(std::chrono::seconds(10));
+  auto status = future.wait_for(std::chrono::seconds(0));
+  AD_CORRECTNESS_CHECK(status != std::future_status::deferred);
+  if (status == std::future_status::timeout) {
+    FAIL() << "Timeout for awaitable reached!";
+  }
+}
```
But why should it return anything else than void? It's just a function that's supposed to run assertions on data, not return anything

RobinTF commented on test/QueryHubTest.cpp:
```diff
@@ -0,0 +1,125 @@
+  std::weak_ptr<QueryToSocketDistributor> observer1;
+  std::weak_ptr<QueryToSocketDistributor> observer2;
+  {
+    auto distributor1 =
+        co_await queryHub.createOrAcquireDistributor(queryId, true);
+    auto distributor2 =
+        co_await queryHub.createOrAcquireDistributor(queryId, true);
+    EXPECT_NE(distributor1, distributor2);
```
See https://github.com/ad-freiburg/qlever/pull/970#discussion_r1333439514

RobinTF commented on test/QueryHubTest.cpp:
```diff
@@ -0,0 +1,125 @@
+        co_await queryHub.createOrAcquireDistributor(queryId, true);
+    EXPECT_EQ(distributor1, distributor2);
+  }
+  EXPECT_NE(distributor1,
+            co_await queryHub.createOrAcquireDistributor(queryId, true));
+  EXPECT_NE(distributor1,
+            co_await queryHub.createOrAcquireDistributor(queryId, false));
```
They would be equal, yes

RobinTF commented on test/QueryHubTest.cpp:
```diff
@@ -0,0 +1,125 @@
+  QueryId queryId = QueryId::idFromString("abc");
+  auto distributor =
+      co_await queryHub.createOrAcquireDistributor(queryId, true);
+  EXPECT_TRUE(co_await distributor->signalStartIfNotStarted());
```
It is automatically called, but setting and checking the start boolean has to happen "atomically" within the strand and I didn't want to add a getter just for testing if the getAndSet function we have here can be used as well. The documentation of `signalStartIfNotStarted` states this also.

RobinTF commented on test/QueryHubTest.cpp:
```diff
@@ -0,0 +1,125 @@
+  QueryId queryId = QueryId::idFromString("abc");
+  auto distributor =
+      co_await queryHub.createOrAcquireDistributor(queryId, false);
+  EXPECT_FALSE(co_await distributor->signalStartIfNotStarted());
```
See https://github.com/ad-freiburg/qlever/pull/970#discussion_r1333502583

RobinTF commented on test/QueryToSocketDistributorTest.cpp:
```diff
@@ -0,0 +1,182 @@
+}
+
+ASYNC_TEST(QueryToSocketDistributor, signalEndRunsCleanup) {
+  bool executed = false;
+  QueryToSocketDistributor queryToSocketDistributor{ioContext,
+                                                    [&]() { executed = true; }};
+  co_await queryToSocketDistributor.signalStartIfNotStarted();
+  EXPECT_FALSE(executed);
+  co_await queryToSocketDistributor.signalEnd();
+  EXPECT_TRUE(executed);
+}
+
+ASYNC_TEST(QueryToSocketDistributor, destructorRunsCleanup) {
+  bool executed = false;
+  {
+    QueryToSocketDistributor queryToSocketDistributor{
+        ioContext, [&]() { executed = true; }};
+    EXPECT_FALSE(executed);
+  }
+  EXPECT_TRUE(executed);
+  // Make sure this is interpreted as a coroutine
+  co_return;
```
This is implicitly tested already by the unique cleanup test, but I get where you're coming from

RobinTF commented on test/QueryToSocketDistributorTest.cpp:
```diff
@@ -0,0 +1,182 @@
+    co_await queryToSocketDistributor.waitForNextDataPiece(0);
+  };
+
+  auto broadcaster = [&]() -> net::awaitable<void> {
+    // Ensure correct order of execution
+    ASSERT_TRUE(waiting);
+    co_await queryToSocketDistributor.signalEnd();
+  };
+  co_await (listener() && broadcaster());
```
I don't think there's a guarantee but that's why I added the assertions and the waiting flags

RobinTF commented on test/QueryToSocketDistributorTest.cpp:
```diff
@@ -0,0 +1,182 @@
+  auto broadcaster = [&]() -> net::awaitable<void> {
+    // Ensure correct order of execution
+    ASSERT_TRUE(waiting);
+    co_await queryToSocketDistributor.signalStartIfNotStarted();
```
Because `addQueryStatusUpdate` sanity checks that it is started before sending anything

RobinTF commented on test/UpdateFetcherTest.cpp:
```diff
@@ -0,0 +1,34 @@
+  QueryHub queryHub{ioContext};
+  UpdateFetcher updateFetcher{queryHub, queryId};
+
+  auto distributor =
+      co_await queryHub.createOrAcquireDistributor(queryId, false);
+  co_await distributor->addQueryStatusUpdate("1");
+  co_await distributor->addQueryStatusUpdate("2");
```
You're right, it should be changed

RobinTF commented on src/util/AsioHelpers.h:
```diff
@@ -0,0 +1,37 @@
+/// Helper function that ensures that co_await resumes on the same
+/// executor it started with.
+template <typename T>
+inline net::awaitable<T> sameExecutor(net::awaitable<T> awaitable) {
+  auto initialExecutor = co_await net::this_coro::executor;
+  T result = co_await std::move(awaitable);
+  co_await net::dispatch(initialExecutor, net::use_awaitable);
+  co_return result;
+}
```
Thinking about this a little bit longer, there's a bit room for discussion there, because this only ever affects catch blocks (and co_spawn entry points if they have a completion token that's not `use_awaitable` or `detached`) and awaitables in general do not guarantee a strand when throwing (as seen by the infinite timer where I manually have to go back to the original strand)

RobinTF commented on test/QueryToSocketDistributorTest.cpp:
```diff
@@ -0,0 +1,182 @@
+
+using ::testing::Pointee;
+
+// Hack to allow ASSERT_*() macros to work with ASYNC_TEST
+#define return co_return
```
It's ugly, but I don't want to define and undefine it for each test case that uses it.

RobinTF commented on test/UpdateWrapperTest.cpp:
```diff
@@ -0,0 +1,70 @@
+  QueryId queryId = QueryId::idFromString("1337");
+  QueryHub queryHub{ioContext};
+
+  auto distributor =
+      co_await queryHub.createOrAcquireDistributor(queryId, false);
```
I'd prefer testing the classes in isolation and I think exposing only a const interface for receiving should be good enough

joka921 commented on src/engine/Server.cpp:
```diff
@@ -617,9 +617,8 @@ boost::asio::awaitable<void> Server::processQuery(
-    auto owningQueryId = getQueryId(request);
-    auto updateWrapper = co_await ad_utility::websocket::UpdateWrapper::create(
-        owningQueryId.toQueryId(), *queryHub_);
+    auto updateWrapper = co_await ad_utility::websocket::MessageSender::create(
+        getQueryId(request), *queryHub_);
```
The variable name should now alwso be `messageSender`.

joka921 commented on src/util/UniqueCleanup.h:
```diff
@@ -47,6 +47,10 @@ class UniqueCleanup {
 
+  /// Runs the cleanup call preemptively by moving out of this instance and
+  /// immediately destructing the new instance
+  void invokeManuallyAndCancel() { auto self = std::move(*this); }
```
This function should either be `&&` qualified, or not move the actual value, else that is really unexpected.
I suggest leaving the value as is and only calling the cleanup and setting `active_` to false.

joka921 commented on src/util/http/websocket/MessageSender.cpp:
```diff
@@ -0,0 +1,46 @@
+
+MessageSender::MessageSender(DistributorWithFixedLifetime distributor,
+                             net::any_io_executor executor)
+    : distributor_{std::move(distributor),
+                   [executor](auto&& distributor) {
+                     auto coroutine =
+                         [](auto distributor) -> net::awaitable<void> {
+                       co_await distributor.distributor_->signalEnd();
+                     };
+                     net::co_spawn(executor, coroutine(AD_FWD(distributor)),
+                                   net::detached);
+                   }},
+      executor_{std::move(executor)} {}
```
I like this solution now, but it has to be explained here:
The distributor is moved into the awaitable, so the `OwningQueryId` which is part of it is only destroyed after the call to `signalEnd` is actually completed and the coroutine is destroyed. That is so implicit that it needs a comment, but again, I like the solution.

joka921 replied:
And also add here that `signalEnd` erases from the queryHub and THEN the owning query Id is unregistered.

joka921 commented on src/util/http/websocket/MessageSender.h:
```diff
@@ -19,30 +19,34 @@ using unique_cleanup::UniqueCleanup;
+class MessageSender {
+ private:
+  /// Keep the OwningQueryId alive until distributor_->signalEnd() is called
+  struct DistributorWithFixedLifetime {
+    std::shared_ptr<QueryToSocketDistributor> distributor_;
+    OwningQueryId owningQueryId_;
+
+    DistributorWithFixedLifetime(
+        std::shared_ptr<QueryToSocketDistributor> distributor,
+        OwningQueryId owningQueryId)
+        : distributor_{std::move(distributor)},
+          owningQueryId_{std::move(owningQueryId)} {}
+  };
```
```suggestion
  struct DistributorWithFixedLifetime {
    std::shared_ptr<QueryToSocketDistributor> distributor_;
    OwningQueryId owningQueryId_;
  };
```
I see no reason why an (implicitly move-only) aggregate shouldn't work here.

joka921 replied:
And rename it to `DistributorAndOwningQueryId`.

RobinTF replied:
> I see no reason why an (implicitly move-only) aggregate shouldn't work here.

It does in theory, but there seems to be a bug in gcc 11 and 12 (didn't bother checking 10), which I mentioned to you a couple of days ago, where this leads to a segmentation fault. I think it might be related to the issue you reported a while ago: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=103909
It also has problems with coroutines and aggregate initialized objects.

joka921 commented on src/util/http/websocket/MessageSender.h:
```diff
@@ -19,30 +19,34 @@ using unique_cleanup::UniqueCleanup;
-  UniqueCleanup<std::shared_ptr<QueryToSocketDistributor>> distributor_;
+class MessageSender {
+ private:
+  /// Keep the OwningQueryId alive until distributor_->signalEnd() is called
```
```suggestion
  /// Keep the OwningQueryId alive until distributor_->signalEnd() is called (see the constructor of `MessageSender` for details).
```

joka921 commented on src/util/http/websocket/QueryHub.cpp:
```diff
@@ -39,38 +34,37 @@ QueryHub::createOrAcquireDistributorInternal(QueryId queryId, bool isProvider) {
-QueryHub::createOrAcquireDistributor(QueryId queryId, bool isProvider) {
-  return sameExecutor(
-      createOrAcquireDistributorInternal(std::move(queryId), isProvider));
+QueryHub::createOrAcquireDistributorForSending(QueryId queryId) {
+  return sameExecutor(createOrAcquireDistributorInternal(std::move(queryId)));
+}
+
+// _____________________________________________________________________________
+
+net::awaitable<std::shared_ptr<const QueryToSocketDistributor>>
+QueryHub::createOrAcquireDistributorForReceiving(QueryId queryId) {
+  auto result = co_await sameExecutor(
+      createOrAcquireDistributorInternal(std::move(queryId)));
+  co_return result;
```
Is this now the semantics of "we might end up with multiple senders/providers for the same queryId, but this is not such a big problem because multiple calls to `signalEnd` are a non-issue because you also check the pointer-equality? Then I understand it now. (The only reason is, that messages might be lost, because the listeners are already unregistered. 

I still think you can keep track of one bool for each queryId on whether we already have picked up the listener, and reintroduce the `isSender` argument to your `...Internal` function.

RobinTF replied:
Multiple calls to `signalEnd()` will raise an assertion error, so that should be clear enough

joka921 commented on src/util/http/websocket/WebSocketSession.h:
```diff
@@ -19,23 +19,24 @@ using boost::asio::ip::tcp;
 
-  /// Loop that waits for an update of the given query to occur and send it
-  /// to the client once it happens
+  /// Wait for updates of the given query and sends them to the client when they
```
```suggestion
  /// Wait for updates of the given query and send them to the client when they
```

joka921 reviewed (changes requested):
The changes get smaller and smaller, we are almost there.

joka921 commented on src/util/UniqueCleanup.h:
```diff
@@ -49,7 +49,12 @@ class UniqueCleanup {
-  void invokeManuallyAndCancel() { auto self = std::move(*this); }
+  void invokeManuallyAndCancel() {
+    if (active_) {
+      std::invoke(function_, std::move(value_));
```
For the same reason as the previous comment, I wouldn't `move` here, but only pass in the reference I think this makes itu usable in more scenarios.

joka921 replied:
```suggestion
      std::invoke(function_, value_);
```

joka921 commented on src/util/http/websocket/MessageSender.cpp:
```diff
@@ -6,12 +6,17 @@
                          [](auto distributor) -> net::awaitable<void> {
+                       // signalEnd() removes the distributor from the QueryHub.
+                       // When the coroutine is destroyed afterwards the query
+                       // id is unregistered from the registry! This is the
```
```suggestion
                       // id is unregistered from the registry by the destructor of `OwningQueryId`! This is the
```

joka921 commented on src/util/http/websocket/QueryHub.cpp:
```diff
@@ -22,14 +22,10 @@ net::awaitable<std::shared_ptr<QueryToSocketDistributor>>
-    // checked for "equality" before removal.
-    socketDistributors_.erase(queryId);
+    auto ptr = socketDistributors_.at(queryId).lock();
+    // Destructor must block until distributor is unregistered
```
```suggestion
    // On all code paths a finished `socketDistributor` is first deleted from the `QueryHub` and only then destroyed. This is asserted by the following check.
```

joka921 commented on src/util/http/websocket/QueryHub.cpp:
```diff
@@ -45,7 +41,23 @@ QueryHub::createOrAcquireDistributorInternal(QueryId queryId) {
   auto distributor = std::make_shared<QueryToSocketDistributor>(
       ioContext_, [this, cleanupAction = std::move(cleanupAction)]() mutable {
-        net::dispatch(globalStrand_, std::move(cleanupAction));
+        auto future =
+            net::dispatch(globalStrand_,
+                          std::packaged_task<void()>(std::move(cleanupAction)));
+        // If we blocked until the cleanup action is executed on different
+        // strand, this could theoretically lead to deadlocks when globalStrand_
+        // is on a different execution context. This is currently not the case,
+        // but it might be in the future, resulting in deadlocks at distance
+        // which is bad. Unfortunately destructors can't be coroutines, which
+        // would just allow us to co_await execution without issues. To prevent
+        // deadlocks, we wait at most 1 second (arbitrarily chosen interval)
+        // for the cleanup to complete, otherwise we throw an exception.
+        auto status = future.wait_for(std::chrono::seconds(1));
+        if (status != std::future_status::ready) {
+          AD_THROW(
+              "Cleanup took longer than 1s to execute, avoiding deadlock"
+              "by throwing an exception instead!");
+        }
```
We'll quickly have to talk about this. 
You should never throw from a Destructor, as it might kill your program if it happens during exception handling. And throwing the exception doesn't help here at all I think, as you still wait for the destructors of all elements.

joka921 reviewed (changes requested):
Very tiny changes and one thing we have to talk about.

joka921 reviewed (approved):
Awesome
